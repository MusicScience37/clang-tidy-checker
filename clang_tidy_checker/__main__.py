"""Main function."""

from clang_tidy_checker.main import main

main()  # pylint: disable=all
