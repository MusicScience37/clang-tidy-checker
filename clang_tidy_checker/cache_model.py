"""Model of cached results."""

import datetime

from sqlalchemy import String
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column

# pylint: disable=too-few-public-methods


class ModelBase(DeclarativeBase):
    """Base class for ORM declarations."""


class CachedCheckResultModel(ModelBase):
    """Table of cached results."""

    __tablename__ = "cached_result"

    source_hash: Mapped[str] = mapped_column(String(100), primary_key=True)
    exit_code: Mapped[int]
    stdout: Mapped[str]
    stderr: Mapped[str]
    created_at: Mapped[datetime.datetime]
