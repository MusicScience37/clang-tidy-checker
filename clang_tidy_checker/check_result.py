"""Results of checks using clang-tidy."""

import dataclasses


@dataclasses.dataclass
class CheckResult:
    """Class of the result of a check."""

    exit_code: int
    stdout: str
    stderr: str
