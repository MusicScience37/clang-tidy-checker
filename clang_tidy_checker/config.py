"""Configuration of this tool.
"""

import dataclasses
import typing

from clang_tidy_checker.search_clang_tidy import search_clang_tidy

# Key of the name of clang-tidy executable to use.
CLANG_TIDY_EXECUTABLE_KEY = "clang_tidy_executable"

# Default name of clang-tidy executable to use.
DEFAULT_CLANG_TIDY_EXECUTABLE = "clang-tidy"

# Key of the path of the build directory.
BUILD_DIR_KEY = "build_dir"

# Default path of the build directory.
DEFAULT_BUILD_DIR = "build"

# Key of the flag to show progress.
SHOW_PROGRESS_KEY = "show_progress"

# Default flag value to show progress.
DEFAULT_SHOW_PROGRESS = True

# Key of checked file patterns.
CHECKED_FILE_PATTERNS_KEY = "file_patterns"

# Default checked files.
DEFAULT_CHECKED_FILE_PATTERNS = [
    "**/*.c",
    "**/*.cpp",
    "**/*.cxx",
    "**/*.cc",
]

# Key of extra arguments.
EXTRA_ARGS_KEY = "extra_args"


@dataclasses.dataclass
class Config:
    """Class of configuration."""

    clang_tidy_path: str
    build_dir: str
    show_progress: bool
    checked_file_patterns: typing.List[str]
    extra_args: typing.List[str]


async def parse_config_from_dict(config: dict) -> Config:
    """Parse configuration from dictionaries.

    Args:
        config (dict): Input dictionary.

    Returns:
        Config: Configuration.
    """

    clang_tidy_path = str(
        config.get(CLANG_TIDY_EXECUTABLE_KEY, DEFAULT_CLANG_TIDY_EXECUTABLE)
    )
    clang_tidy_path = await search_clang_tidy(clang_tidy_path)

    build_dir = str(config.get(BUILD_DIR_KEY, DEFAULT_BUILD_DIR))

    show_progress = bool(config.get(SHOW_PROGRESS_KEY, DEFAULT_SHOW_PROGRESS))

    checked_file_patterns = list(
        config.get(CHECKED_FILE_PATTERNS_KEY, DEFAULT_CHECKED_FILE_PATTERNS)
    )

    extra_args = [str(elem) for elem in config.get(EXTRA_ARGS_KEY, [])]

    return Config(
        clang_tidy_path=clang_tidy_path,
        build_dir=build_dir,
        show_progress=show_progress,
        checked_file_patterns=checked_file_patterns,
        extra_args=extra_args,
    )
