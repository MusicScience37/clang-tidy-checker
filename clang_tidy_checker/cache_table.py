"""Tables of cached results."""

import datetime
import typing

import sqlalchemy
import sqlalchemy.orm

from clang_tidy_checker.cache_model import CachedCheckResultModel, ModelBase
from clang_tidy_checker.check_result import CheckResult


class CacheTable:
    """Tables of cached results."""

    def __init__(self, engine: sqlalchemy.Engine, max_cache_entries: int) -> None:
        self._engine = engine
        ModelBase.metadata.create_all(engine)
        self._max_cache_entries = max_cache_entries

    def save(self, source_hash: str, result: CheckResult) -> None:
        """Save a result of a check."""
        with sqlalchemy.orm.Session(self._engine) as session:
            cached_result = CachedCheckResultModel(
                source_hash=source_hash,
                exit_code=result.exit_code,
                stdout=result.stdout,
                stderr=result.stderr,
                created_at=datetime.datetime.now(),
            )
            session.add(cached_result)
            session.commit()

        self._remove_old_entries()

    def load(self, source_hash: str) -> typing.Optional[CheckResult]:
        """Load a cached result of a check.

        Args:
            source_hash (str): Hash of source code.

        Returns:
            typing.Optional[CheckResult]: Cached result if found.
        """
        with sqlalchemy.orm.Session(self._engine) as session:
            statement = sqlalchemy.select(CachedCheckResultModel).where(
                CachedCheckResultModel.source_hash == source_hash
            )
            cached_result = session.scalars(statement).one_or_none()
            if cached_result is None:
                return None
            return CheckResult(
                exit_code=cached_result.exit_code,
                stdout=cached_result.stdout,
                stderr=cached_result.stderr,
            )

    def _remove_old_entries(self) -> None:
        """Remove old entries."""
        with sqlalchemy.orm.Session(self._engine) as session:
            current_num_entries = typing.cast(
                int,
                session.scalar(
                    sqlalchemy.select(
                        # Pylint wrongly generate an error.
                        # pylint: disable=not-callable
                        sqlalchemy.func.count(CachedCheckResultModel.source_hash)
                    )
                ),
            )
            if current_num_entries <= self._max_cache_entries:
                return
            removed_entries = current_num_entries - self._max_cache_entries
            statement = (
                sqlalchemy.select(CachedCheckResultModel)
                .order_by(CachedCheckResultModel.created_at.asc())
                .limit(removed_entries)
            )
            for entry in session.scalars(statement):
                session.delete(entry)
            session.commit()


def create_cache_table_at(filepath: str, max_cache_entries: int) -> CacheTable:
    """Create a table of caches at a file path.

    Args:
        filepath (str): File path.
        max_cache_entries (int): Maximum number of entries in the cache.

    Returns:
        CacheTable: Created table.
    """
    return CacheTable(
        engine=sqlalchemy.create_engine(f"sqlite:///{filepath}"),
        max_cache_entries=max_cache_entries,
    )
