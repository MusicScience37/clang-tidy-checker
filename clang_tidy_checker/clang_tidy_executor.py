"""Class to execute clang-tidy."""

import abc
import dataclasses
import logging
import os
import typing

import msgpack

from clang_tidy_checker.command_executor import CommandExecutor
from clang_tidy_checker.config import Config
from clang_tidy_checker.source_hash_calculator import SourceHashCalculator

try:
    from typing import Self
except ImportError:
    Self = typing.TypeVar("Self", bound="IClangTidyExecutor")  # type: ignore

LOGGER = logging.getLogger(__name__)


@dataclasses.dataclass
class CheckResult:
    """Class of the result of a check."""

    exit_code: int
    stdout: str
    stderr: str


class IClangTidyExecutor(abc.ABC):
    """Interface for clang-tidy executor."""

    @abc.abstractmethod
    async def __aenter__(self) -> Self:
        pass

    @abc.abstractmethod
    async def __aexit__(self, exc_type, exc_value, traceback) -> None:
        pass

    @abc.abstractmethod
    async def execute(self, *, input_file: str) -> CheckResult:
        """Execute clang-tidy.

        Args:
            input_file (str): Input file path.

        Returns:
            CheckResult: Result.
        """


def write_result_log(input_file: str, exit_code: int, stdout: str, stderr: str) -> None:
    """Write log of the result.

    Args:
        input_file (str): Input file path.
        exit_code (int): Exit code.
        stdout (str): Standard output.
        stderr (str): Standard error.
    """

    if exit_code == 0:
        LOGGER.info("Check of %s finished with exit code %s.", input_file, exit_code)
        if stdout != "":
            LOGGER.debug("%s", stdout)
        if stderr != "":
            LOGGER.debug("%s", stderr)
    else:
        LOGGER.warning(
            "Check of %s finished with exit code %s.\n%s\n%s",
            input_file,
            exit_code,
            stdout,
            stderr,
        )


class ClangTidyExecutor(IClangTidyExecutor):
    """Class to execute clang-tidy."""

    def __init__(self, config: Config) -> None:
        self._config = config
        self._command_executor = CommandExecutor()

    async def __aenter__(self) -> Self:
        await self._command_executor.__aenter__()
        return self

    async def __aexit__(self, exc_type, exc_value, traceback) -> None:
        await self._command_executor.__aexit__(exc_type, exc_value, traceback)

    async def execute(self, *, input_file: str) -> CheckResult:
        command = (
            [
                self._config.clang_tidy_path,
                "--quiet",
                "--warnings-as-errors=*",
                "-p",
                self._config.build_dir,
            ]
            + self._config.extra_args
            + [
                input_file,
            ]
        )

        result = await self._command_executor.execute(
            command=command, cwd=self._config.build_dir
        )
        exit_code = result.exit_code
        stdout = result.stdout
        stderr = result.stderr

        write_result_log(
            exit_code=exit_code,
            input_file=input_file,
            stdout=stdout,
            stderr=stderr,
        )

        return CheckResult(exit_code=exit_code, stdout=stdout, stderr=stderr)


def get_cache_file_path(cache_dir: str, source_hash: str) -> str:
    """Get the file path of cache file.

    Args:
        cache_dir (str): Path to the cache directory.
        source_hash (str): Hash of the source code.

    Returns:
        str: Path to the cache file.
    """
    return os.path.join(
        cache_dir, source_hash[-1], source_hash[-2], source_hash[-20:-2]
    )


class CachedClangTidyExecutor(IClangTidyExecutor):
    """Class to execute clang-tidy but with caching of results."""

    def __init__(self, config: Config) -> None:
        self._clang_tidy_executor = ClangTidyExecutor(config=config)
        self._source_hash_calculator = SourceHashCalculator(config=config)
        if config.cache_dir is None:
            raise ValueError("Cache directory is required for CachedClangTidyExecutor.")
        self._cache_dir = config.cache_dir

    async def __aenter__(self) -> Self:
        await self._clang_tidy_executor.__aenter__()
        await self._source_hash_calculator.__aenter__()
        return self

    async def __aexit__(self, exc_type, exc_value, traceback) -> None:
        await self._clang_tidy_executor.__aexit__(exc_type, exc_value, traceback)
        await self._source_hash_calculator.__aexit__(exc_type, exc_value, traceback)

    async def execute(self, *, input_file: str) -> CheckResult:
        source_hash = await self._source_hash_calculator.calculate(
            input_file=input_file
        )
        cache_file_path = get_cache_file_path(
            cache_dir=self._cache_dir, source_hash=source_hash
        )
        if os.path.exists(cache_file_path):
            with open(cache_file_path, "rb") as file:
                result_dict = msgpack.unpack(file)
            exit_code = int(result_dict["error_code"])
            stdout = str(result_dict["stdout"])
            stderr = str(result_dict["stderr"])

            write_result_log(
                exit_code=exit_code,
                input_file=input_file,
                stdout=stdout,
                stderr=stderr,
            )

            return CheckResult(exit_code=exit_code, stdout=stdout, stderr=stderr)

        result = await self._clang_tidy_executor.execute(input_file=input_file)
        os.makedirs(os.path.dirname(cache_file_path), exist_ok=True)
        with open(cache_file_path, "wb") as file:
            msgpack.pack(
                {
                    "error_code": result.exit_code,
                    "stdout": result.stdout,
                    "stderr": result.stderr,
                },
                file,
            )
        return result
