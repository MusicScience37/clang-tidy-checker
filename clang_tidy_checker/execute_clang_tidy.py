"""Execute clang-tidy."""

import dataclasses
import logging

import trio

from clang_tidy_checker.config import Config

LOGGER = logging.getLogger(__name__)


@dataclasses.dataclass
class ExecutionResult:
    """Class of the result of an execution."""

    input_file: str
    exit_code: int
    stdout: str
    stderr: str


# TODO: better handling.
CONSOLE_ENCODING = "utf8"


async def execute_clang_tidy(*, config: Config, input_file: str) -> ExecutionResult:
    """Execute clang-tidy.

    Args:
        config (Config): Configuration.
        input_file (str): Input file path.

    Returns:
        ExecutionResult: Result of execution.
    """

    result = await trio.run_process(
        [
            config.clang_tidy_path,
            "--quiet",
            "--warnings-as-errors=*",
            "-p",
            config.build_dir,
        ]
        + config.extra_args
        + [
            input_file,
        ],
        capture_stdout=True,
        capture_stderr=True,
        check=False,
    )
    stdout = result.stdout.decode(encoding=CONSOLE_ENCODING)
    stderr = result.stderr.decode(encoding=CONSOLE_ENCODING)

    if result.returncode == 0:
        LOGGER.info(
            "Check of %s finished with exit code %s.", input_file, result.returncode
        )
        if result.stdout != "":
            LOGGER.debug("%s", stdout)
        if result.stderr != "":
            LOGGER.debug("%s", stderr)
    else:
        LOGGER.warning(
            "Check of %s finished with exit code %s.\n%s\n%s",
            input_file,
            result.returncode,
            stdout,
            stderr,
        )

    return ExecutionResult(
        input_file=input_file,
        exit_code=result.returncode,
        stdout=stdout,
        stderr=stderr,
    )
