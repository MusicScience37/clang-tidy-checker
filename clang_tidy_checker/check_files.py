"""Check files.
"""

import typing

import tqdm
import tqdm.contrib.logging

from clang_tidy_checker.config import Config
from clang_tidy_checker.execute_clang_tidy import execute_clang_tidy


async def check_files(*, config: Config, input_files: typing.List[str]) -> bool:
    """Check files.

    Args:
        config (Config): Configuration.
        input_files (typing.List[str]): Files.

    Returns:
        bool: True if no error, False otherwise.
    """

    has_error = False

    with tqdm.contrib.logging.logging_redirect_tqdm():
        tqdm_obj = tqdm.tqdm(
            total=len(input_files), unit="file", disable=not config.show_progress
        )
        # TODO: multiple files at once.
        for input_file in input_files:
            result = await execute_clang_tidy(config=config, input_file=input_file)
            if result.exit_code != 0:
                has_error = True
            tqdm_obj.update()
        tqdm_obj.close()

    return not has_error
