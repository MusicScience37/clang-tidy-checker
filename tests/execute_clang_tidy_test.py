"""Test of execute_clang_tidy.py.
"""

import copy
import pathlib

import approvaltests
import approvaltests.scrubbers
import trio
import trio.testing

from clang_tidy_checker.config import Config
from clang_tidy_checker.execute_clang_tidy import execute_clang_tidy, ExecutionResult

from .path_scrubber import PATH_SCRUBBER
from .warning_count_scrubber import WARNING_COUNT_SCRUBBER


def check_result(result: ExecutionResult):
    """Check a result.

    Args:
        result (ExecutionResult): Result.
    """

    approvaltests.approvals.verify(
        f"""input_file: {result.input_file}
exit_code: {result.exit_code}
stdout:
{result.stdout}
stderr:
{result.stderr}
""",
        options=approvaltests.Options().with_scrubber(
            approvaltests.scrubbers.combine_scrubbers(
                PATH_SCRUBBER, WARNING_COUNT_SCRUBBER
            )
        ),
    )


@trio.testing.trio_test
async def test_execute_clang_tidy_no_error(
    default_config: Config, sample_proj_no_error: pathlib.Path
):
    """Test of execute_clang_tidy resulting in no error."""

    config = copy.deepcopy(default_config)
    config.build_dir = str(sample_proj_no_error / "build")
    input_file = str(sample_proj_no_error / "src" / "sample_function.cpp")

    result = await execute_clang_tidy(config=config, input_file=input_file)

    check_result(result)


@trio.testing.trio_test
async def test_execute_clang_tidy_warning(
    default_config: Config, sample_proj_warning: pathlib.Path
):
    """Test of execute_clang_tidy resulting in a warning."""

    config = copy.deepcopy(default_config)
    config.build_dir = str(sample_proj_warning / "build")
    input_file = str(sample_proj_warning / "src" / "sample_function.cpp")

    result = await execute_clang_tidy(config=config, input_file=input_file)

    check_result(result)


@trio.testing.trio_test
async def test_execute_clang_tidy_error(
    default_config: Config, sample_proj_error: pathlib.Path
):
    """Test of execute_clang_tidy resulting in an error."""

    config = copy.deepcopy(default_config)
    config.build_dir = str(sample_proj_error / "build")
    input_file = str(sample_proj_error / "src" / "sample_function.cpp")

    result = await execute_clang_tidy(config=config, input_file=input_file)

    check_result(result)
