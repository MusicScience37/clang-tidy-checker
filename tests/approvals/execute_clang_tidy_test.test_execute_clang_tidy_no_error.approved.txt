input_file: <root-dir>/sample_inputs/proj_no_error/src/sample_function.cpp
exit_code: 0
stdout:

stderr:
37323 warnings generated.

